### PR TITLE
run_loop for mask rcnn first commit

### PR DESCRIPTION
**Pull request by MehdiShz**
It will alternate between 1 epoch of training and one evaluation

### CODE REVIEW

File: object_detection/Mask_RCNN/mask_rcnn_run_loop.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+  pass
+
+
+def evaluate(_):
+  # TODO(mehdi): evaluating with mask and box metrics
+  pass
+
+
+def main(_):
+  assert FLAGS.train_dir, '`train_dir` is missing.'
```
Comment (k-w-w) [inline object_detection/Mask_RCNN/mask_rcnn_run_loop.py]: You can set the flag default to `None` and use `mark_flag_as_required to mark required flags. For example:
```
flags.DEFINE_string('train_dir', None, '...')
flags.mark_flag_as_required('train_dir')
```

File: object_detection/Mask_RCNN/mask_rcnn_run_loop.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+                                     is_training=True)
+  eval_model_fn = functools.partial(model_builder.build,
+                                    model_config=model_config,
+                                    is_training=False)
+
+  def get_next(config):
+    return dataset_util.make_initializable_iterator(
+        dataset_builder.build(config)).get_next()
+
+  train_input_dict_fn = functools.partial(get_next, train_input_config)
```
Comment (k-w-w) [inline object_detection/Mask_RCNN/mask_rcnn_run_loop.py]: A comment here about train_input_dict_fn and eval_input_dict would be helpful. (In particular - explaining the use of `func.partial(get_next,...)`)?

File: object_detection/Mask_RCNN/mask_rcnn_run_loop.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+  eval_input_dict_fn = functools.partial(get_next, eval_input_config)
+
+  env = json.loads(os.environ.get('TF_CONFIG', '{}'))
+  cluster_data = env.get('cluster', None)
+  cluster = tf.train.ClusterSpec(cluster_data) if cluster_data else None
+  task_data = env.get('task', None) or {'type': 'master', 'index': 0}
```
Comment (k-w-w) [inline object_detection/Mask_RCNN/mask_rcnn_run_loop.py]: Might be cleaner to do `task_data =  env.get('task', {'type': 'master', 'index': 0})`

File: object_detection/Mask_RCNN/mask_rcnn_run_loop.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+  task_info = type('TaskSpec', (object,), task_data)
+
+  # Parameters for a single worker.
```
Comment (k-w-w) [inline object_detection/Mask_RCNN/mask_rcnn_run_loop.py]: Should we use the flag values here?

File: object_detection/Mask_RCNN/mask_rcnn_run_loop.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+def evaluate(_):
+  # TODO(mehdi): evaluating with mask and box metrics
+  pass
+
+
+def main(_):
+  assert FLAGS.train_dir, '`train_dir` is missing.'
+  assert FLAGS.pipeline_config_path, '`pipeline_config_path` is missing'
+  assert FLAGS.checkpoint_dir, '`checkpoint_dir` is missing.'
```
Comment (k-w-w) [inline object_detection/Mask_RCNN/mask_rcnn_run_loop.py]: This flag has been commented out. Might be better to remove completely

File: object_detection/Mask_RCNN/mask_rcnn_run_loop.py (hunk 2 above)
Comment (MehdiShz) [reply object_detection/Mask_RCNN/mask_rcnn_run_loop.py]: changed to
train_input_dict_fn = get_next(train_input_config)
as the dataset_builder doesn't need another argument

File: object_detection/Mask_RCNN/mask_rcnn_run_loop.py (hunk 4 above)
Comment (MehdiShz) [reply object_detection/Mask_RCNN/mask_rcnn_run_loop.py]: FLAG values will be passed to trainer.train later

File: object_detection/Mask_RCNN/mask_rcnn_run_loop.py (hunk 1 above)
Comment (MehdiShz) [reply object_detection/Mask_RCNN/mask_rcnn_run_loop.py]: why? does it work as the "assert FLAGS. ..."

File: object_detection/Mask_RCNN/configs/mask_rcnn_resnet50_atrous_coco.config [hunk 6]
```diff
@@ -0,0 +1,169 @@
+# Mask R-CNN with Resnet-50 (v1), Atrous version
```
Comment (nealwu) [inline object_detection/Mask_RCNN/configs/mask_rcnn_resnet50_atrous_coco.config:1]: This file is just from https://github.com/tensorflow/models/blob/master/research/object_detection/samples/configs/mask_rcnn_resnet50_atrous_coco.config right?

File: object_detection/Mask_RCNN/mask_rcnn_run_loop.py [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+from object_detection.builders import dataset_builder
+from object_detection.builders import graph_rewriter_builder
+from object_detection.builders import model_builder
+from object_detection.utils import config_util
+from object_detection.utils import dataset_util
+from object_detection.utils import label_map_util
+
+import tensorflow as tf
+
+EPOCHS_BETWEEN_EVALS = 1
```
Comment (nealwu) [inline object_detection/Mask_RCNN/mask_rcnn_run_loop.py]: We should go ahead and make this a flag as well.

File: object_detection/Mask_RCNN/mask_rcnn_run_loop.py [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+  assert mask_min_ap == -1 or 0 < mask_min_ap < 1
+  assert box_min_ap == -1 or 0 < box_min_ap < 1
+  try:
+    mask_mAP_reached = eval_metrics['DetectionMasks_Precision/mAP']
+    box_mAP_reached = eval_metrics['DetectionBoxes_Precision/mAP']
+  except KeyError as err:
+    raise Exception('eval_metrics dict does not contain the mAP field') from err
+
+  return (mask_min_ap == -1 or mask_mAP_reached > mask_min_ap) & \
+         (box_min_ap == -1 or box_mAP_reached > box_min_ap)
```
Comment (nealwu) [inline object_detection/Mask_RCNN/mask_rcnn_run_loop.py]: nit: the `&` should be `and` instead.

More importantly, doesn't this return true when both parameters are left unset?

File: object_detection/Mask_RCNN/mask_rcnn_run_loop.py [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+
+  if train_config.num_steps:
+    total_num_epochs = train_config.num_steps
+    train_config.num_steps = EPOCHS_BETWEEN_EVALS
+    total_training_cycle = total_num_epochs//train_config.num_steps
+  else:
+    # TODO(mehdi): make it run indef
+    total_num_epochs = 200000
+    train_config.num_steps = EPOCHS_BETWEEN_EVALS
+    total_training_cycle = total_num_epochs//train_config.num_steps
```
Comment (nealwu) [inline object_detection/Mask_RCNN/mask_rcnn_run_loop.py]: nit: add spaces around `//`

File: object_detection/Mask_RCNN/mask_rcnn_run_loop.py [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+flags.DEFINE_string('pipeline_config_path', '',
+                    'Path to a pipeline_pb2.TrainEvalPipelineConfig config '
+                    'file. If provided, other configs are ignored')
+
+flags.DEFINE_boolean('eval_training_data', False,
+                     'If training data should be evaluated for this job.')
+# check point dir is the same as the train dir
+# flags.DEFINE_string('checkpoint_dir', '',
+#                    'Directory containing checkpoints to evaluate, typically '
+#                    'set to `train_dir` used in the training job.')
```
Comment (nealwu) [inline object_detection/Mask_RCNN/mask_rcnn_run_loop.py]: If we're not using this for now let's just remove it.

File: object_detection/Mask_RCNN/mask_rcnn_run_loop.py [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+    raise Exception('eval_metrics dict does not contain the mAP field') from err
+
+  return (mask_min_ap == -1 or mask_mAP_reached > mask_min_ap) & \
+         (box_min_ap == -1 or box_mAP_reached > box_min_ap)
+
+
+def main(_):
+  assert FLAGS.train_dir, '`train_dir` is missing.'
+  assert FLAGS.pipeline_config_path, '`pipeline_config_path` is missing'
+  assert FLAGS.eval_dir, '`eval_dir` is missing.'
```
Comment (nealwu) [inline object_detection/Mask_RCNN/mask_rcnn_run_loop.py:107]: @k-w-w 's [suggestion](https://github.com/MehdiShz/reference/pull/2#discussion_r197592005) to use `mark_flag_as_required` seems like a preferable pattern here. (For some reason I can no longer reply to that comment.)

File: object_detection/Mask_RCNN/mask_rcnn_run_loop.py [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+
+  flags.DEFINE_string('checkpoint_dir', FLAGS.train_dir,
+                      'directory to read checkpoints from '
+                      'which is the same as the train_dir')
```
Comment (nealwu) [inline object_detection/Mask_RCNN/mask_rcnn_run_loop.py]: Better not to define a flag here separately from all the other flags. If `checkpoint_dir` is almost always the same as `train_dir`, may as well just use `train_dir` instead.

File: object_detection/Mask_RCNN/mask_rcnn_run_loop.py [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+  master = ''
+
+  if cluster_data and 'worker' in cluster_data:
+    # Number of total worker replicas include "worker"s and the "master".
+    worker_replicas = len(cluster_data['worker']) + 1
+  if cluster_data and 'ps' in cluster_data:
+    ps_tasks = len(cluster_data['ps'])
+
+  if worker_replicas > 1 and ps_tasks < 1:
+    raise ValueError('At least 1 ps task is needed for distributed training.')
```
Comment (nealwu) [inline object_detection/Mask_RCNN/mask_rcnn_run_loop.py]: Let's spell out 'ps' here as 'parameter server'

**Review by nealwu — COMMENTED** [review]
Looks reasonable so far. Let us know when you have a better idea of the training progress / metrics.

File: object_detection/Mask_RCNN/configs/mask_rcnn_resnet50_atrous_coco.config (hunk 6 above)
Comment (MehdiShz) [reply object_detection/Mask_RCNN/configs/mask_rcnn_resnet50_atrous_coco.config:1]: yes
